src/Core/Framework/JWT/SalesChannel/JWTGenerator.php

```diff
@@ -0,0 +1,151 @@
+     *
+     * @return non-empty-string
+     */
+    public function encode(JWTStruct $jwt): string
+    {
+        return $this->buildToken($jwt)->getToken(
+            $this->configuration->signer(),
+            $this->configuration->signingKey()
+        )->toString();
+    }
+
+    /**
+     * @return T
+     */
+    public function decode(string $token): JWTStruct
+    {
+        if (!$token) {
+            throw JWTException::invalidJwt('JWT cannot be empty');
+        }
+
+        $jwt = $this->configuration->parser()->parse($token);
+        if (!$jwt instanceof UnencryptedToken) {
+            throw JWTException::invalidJwt('JWT is not an unencrypted token');
+        }
+
+        if (!$this->configuration->validator()->validate($jwt, ...$this->getValidationConstraints())) {
+            throw JWTException::invalidJwt('JWT validation failed');
+        }
+
+        $structClass = $this->getJWTStructClass();
+        $claims = $jwt->claims()->all();
+        $this->validator->validate($claims, $this->getConstraints());
+
+        return new ($structClass)($claims);
+    }
+
+    /**
+     * @param T $jwt
+     */
+    protected function buildToken(JWTStruct $jwt): Builder
+    {
+        $now = new \DateTimeImmutable('@' . time());
+
+        $builder = $this->configuration->builder()
+            ->issuedAt($jwt->iat ?? $now)
+            ->canOnlyBeUsedAfter($jwt->nbf ?? $now)
+            ->expiresAt($jwt->exp ?? $now->modify(\sprintf('+%d seconds', $this->getTokenLifetime())));
+
+        foreach ($jwt->getVars() as $key => $value) {
+            if (!$value && \in_array($key, RegisteredClaims::ALL, true) || \in_array($key, RegisteredClaims::DATE_CLAIMS, true)) {
+                continue;
+            }
+
+            if ($key === RegisteredClaims::ID) {
+                $builder = $builder->identifiedBy($value ?: Uuid::randomHex());
+                continue;
+            }
+
+            if (!$key || !$value) {
+                continue;
+            }
+
+            if ($key === RegisteredClaims::AUDIENCE) {
+                $builder = $builder->permittedFor(...$value);
+                continue;
+            }
+
+            if ($key === RegisteredClaims::SUBJECT) {
+                $builder = $builder->relatedTo($value);
+                continue;
+            }
+
+            if ($key === RegisteredClaims::ISSUER) {
+                $builder = $builder->issuedBy($value);
+                continue;
+            }
+
+            $builder = $builder->withClaim($key, $value);
+        }
+
+        return $builder;
+    }
+
+    /**
+     * @return array<Constraint>
+     */
+    protected function getValidationConstraints(): array
+    {
+        return $this->configuration->validationConstraints();
+    }
+
+    /**
+     * lifetime in seconds (default is 1h)
+     */
+    protected function getTokenLifetime(): int
+    {
+        return 3600;
+    }
+
+    protected function getConstraints(): DataValidationDefinition
```

Comment:
I find the naming `getValidationConstraints` and `getConstraints` irritating, both are used for validation

Reply:
Are there even any applications for changing the configuration's constraints (`getValidationConstraints`)?

Reply:
Originally, I prepared this method for refactoring the JWT of IAP into this, but then I realized that this does not adhere to the norm (AKA the payload being an array/a Collection instead of struct), so this won't be used there.